### PR TITLE
Clean up the class generator.

### PR DESCRIPTION
**Pull request by mbullington**
I didn't break rendering into their own methods (yet) because of the large amount of variables that needed to be passed as parameters to each one. That may change in the future.

### CODE REVIEW

File: lib/src/compiler/nodes/class.dart [hunk 1]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
+          .map((f) => compiler.info.getElement(f.split("/")[0], f.split("/")[1]))
+          .toList();
 
-        var type = child[0];
-        var id = child[1];
-
-        var data = compiler.info.getElement(type, id);
-        var name = data["name"];
+      int index = 0;
+      for (final _ in memberData["children"]) {
```
Comment (Pacane) [inline lib/src/compiler/nodes/class.dart]: this could be done with a regular for

File: lib/src/compiler/nodes/class.dart [hunk 2]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
+              final String constructorName = name.substring(className.length + 1);
+              
+              buf.write("""
+                mdex.$className.$constructorName = function() {
+                  var classObj = Object.create(mdex.$className.prototype);
+                  classObj[$symDartObj] = 
+              """);
+            }
+
+            final code = data["code"] == null || data["code"].length == 0
```
Comment (Pacane) [inline lib/src/compiler/nodes/class.dart]: .isEmpty

File: lib/src/compiler/nodes/class.dart [hunk 3]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
-              data,
-              compiler.analyzer.getFunctionParameters(
-                  c.libraryName, data["name"], memberData["name"]));
-
-          if (c != null &&
-              c.getters.contains(data["name"]) &&
+          // defer handling of getters/setters
+          
+          if (c != null && c.getters.contains(data["name"]) &&
               params.length == 0) {
```
Comment (Pacane) [inline lib/src/compiler/nodes/class.dart]: isEmpty

File: lib/src/compiler/nodes/class.dart [hunk 4]
```diff
@@ -128,96 +192,89 @@ class Class implements Renderable {
 
           if (data["code"].length > 0) {
```
Comment (Pacane) [inline lib/src/compiler/nodes/class.dart]: isNotEmpty

File: lib/src/compiler/nodes/class.dart [hunk 5]
```diff
@@ -128,96 +192,89 @@ class Class implements Renderable {
...
+          if (names.contains(name) || name.startsWith("_")) continue;
+          names.add(name);
+
+          // TODO: static fields
+          if (c == null || c.staticFields.contains(name)) continue;
+          
+          final codeParts = data["code"]
+              .split("\n")
+              .where((name) =>
+                  name.length > 0 &&
```
Comment (Pacane) [inline lib/src/compiler/nodes/class.dart]: isNotEmpty

File: lib/src/compiler/nodes/class.dart [hunk 6]
```diff
@@ -128,96 +192,89 @@ class Class implements Renderable {
...
+                  !name.contains(" ") &&
+                  name.contains(_FIELD_REGEX))
+              .map((name) {
+                try {
+                  if (name.contains(":"))
+                    name = name.substring(name.indexOf(":") + 1);
+                  return _FIELD_REGEX.firstMatch(name).group(0);
+                } catch (e) {
+                  throw name;
```
Comment (Pacane) [inline lib/src/compiler/nodes/class.dart]: Give better message here.

File: lib/src/compiler/nodes/class.dart [hunk 7]
```diff
@@ -242,20 +300,29 @@ class Class implements Renderable {
 
         if (setters[accessor] != null) {
           prototype.write("set $accessor(v) {");
```
Comment (Pacane) [inline lib/src/compiler/nodes/class.dart]: I'd rename `v` to something more explicit.

File: lib/compiler.dart [hunk 8]
```diff
@@ -69,23 +69,52 @@ final String _OBJ_EACH_PREFIX = """
...
+  // a reference to the javascript instance wrapping the dart object that's
+  // put on the object when it's instated.
+  
+  // see the base transformer, it uses this symbol to return the actual JS instance
+  // instead of creating a new wrapper the class.
+  static const String jsObj = "clId";
+  String get symJsObj => _SymbolTypes.jsObj;
+}
+
 final String _OVERRIDE_PREFIX = """
```
Comment (kaendfinger) [inline lib/compiler.dart]: _overridePrefix

File: lib/compiler.dart [hunk 9]
```diff
@@ -69,23 +69,52 @@ final String _OBJ_EACH_PREFIX = """
...
+    
+  // backup object for when methods are overwritten
+  // when a method is overwritten, the original method is pushed into the
+  // backup object.
+  
+  // then when a super call is made from human-readable JS, the backup object
+  // is called instead of the actual object, so the super call does not end
+  // in a recursive loop. 
+  static const String backup = "clBk";
+  String get symBackup => _SymbolTypes.backup;
```
Comment (kaendfinger) [inline lib/compiler.dart]: symbolBackup

File: lib/compiler.dart [hunk 10]
```diff
@@ -69,23 +69,52 @@ final String _OBJ_EACH_PREFIX = """
...
 
+abstract class _SymbolTypes {
+  // an identifier in the prototype of a calzone wrapped class to indicate that
+  // it also has the below two symbols
+  static const String isWrapped = "clIw";
+  String get symIsWrapped => _SymbolTypes.isWrapped;
+  
+  // a reference to the dart object that the calzone class instance is wrapping
+  static const String dartObj = "clOb";
+  String get symDartObj => _SymbolTypes.dartObj;
```
Comment (kaendfinger) [inline lib/compiler.dart]: symbolDartObject or symbolObject

File: lib/compiler.dart [hunk 11]
```diff
@@ -69,23 +69,52 @@ final String _OBJ_EACH_PREFIX = """
   }
 """;
 
+abstract class _SymbolTypes {
+  // an identifier in the prototype of a calzone wrapped class to indicate that
+  // it also has the below two symbols
+  static const String isWrapped = "clIw";
+  String get symIsWrapped => _SymbolTypes.isWrapped;
```
Comment (kaendfinger) [inline lib/compiler.dart]: symbolIsWrapped (I and L are confusing, so maybe not this)

File: lib/src/compiler/nodes/class.dart [hunk 12]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
+        final data = childData[index++];
+        
+        final type = data["kind"];
+        String name = data["name"];
 
         if (type == "function") {
           if (names.contains(name) || name.startsWith("_")) continue;
 
+          // name replacements for operator overloading functions
           if (NAME_REPLACEMENTS.containsKey(data["name"])) {
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: NAME_REPLACEMENTS => replacementNames or nameReplacements (Dart Style Guide recommends this)

File: lib/src/compiler/nodes/class.dart [hunk 13]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
-                .map((f) => compiler.info
-                    .getElement(f.split("/")[0], f.split("/")[1])["name"])
-                .contains(NAME_REPLACEMENTS[data["name"]])) continue;
-            name = NAME_REPLACEMENTS[data["name"]];
-            if (names.contains(name)) continue;
+            var nameReplacement = NAME_REPLACEMENTS[data["name"]];
+            
+            if (names.contains(nameReplacement) ||
+                childData.map((data) => data["name"]).contains(nameReplacement))
+              continue;
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Braces around this.

File: lib/src/compiler/nodes/class.dart [hunk 14]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
+          if (isConstructor && !isTopLevel) continue;
+          
+          final params = _getParamsFromInfo(compiler, data,
+              compiler.analyzer.getFunctionParameters(
+                  c.libraryName, isConstructor ? this.data["name"] : data["name"],
+                  memberData["name"]));
+
+          // if unnamed or named constructor
+          if (isConstructor) {
+            var buf = isDefaultConstructor ? constructor : global;
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: buf => buffer or buff

File: lib/src/compiler/nodes/class.dart [hunk 15]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
             buf.write(".apply(this, arguments);");
-            if (isDefault)
-              buf.write("this[clOb][clBk] = {};this[clOb][clId] = this;");
-            else {
-              buf.write(
-                  "classObj[clOb][clBk] = {};classObj[clOb][clId] = classObj;return classObj;};");
-            }
+            
+            if (isDefaultConstructor)
+              continue;
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Braces

File: lib/src/compiler/nodes/class.dart [hunk 16]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
-              compiler.analyzer.getFunctionParameters(
-                  c.libraryName, data["name"], memberData["name"]));
-
-          if (c != null &&
-              c.getters.contains(data["name"]) &&
+          // defer handling of getters/setters
+          
+          if (c != null && c.getters.contains(data["name"]) &&
               params.length == 0) {
             if (!accessors.contains(name)) accessors.add(name);
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Braces

File: lib/src/compiler/nodes/class.dart [hunk 17]
```diff
@@ -2,124 +2,188 @@ part of calzone.compiler;
...
             getters[name] = data;
             continue;
           }
 
-          if (c != null &&
-              c.setters.contains(data["name"]) &&
+          if (c != null && c.setters.contains(data["name"]) &&
               params.length == 1) {
             if (!accessors.contains(name)) accessors.add(name);
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Braces

File: lib/src/compiler/nodes/class.dart [hunk 18]
```diff
@@ -128,96 +192,89 @@ class Class implements Renderable {
...
-                        binding: "this[clOb]",
+                        binding: "this[$symDartObj]",
                         prefix: "mdex.${this.data["name"]}.prototype",
-                        code: "this[clOb].${data["code"].split(":")[0]}"))
+                        code: "this[$symDartObj].${data["code"].split(":")[0]}"))
                     .render(compiler, global);
               } else {
-                prototype.write(data["name"] + ": ");
+                prototype.write("$name: ");
                 (new Func(data, params,
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Assign this to a variable then call it, that's generally a good rule.

File: lib/src/compiler/nodes/class.dart [hunk 19]
```diff
@@ -128,96 +192,89 @@ class Class implements Renderable {
...
-
-            prototype.write("set ${data["name"]}(v) {");
-            compiler.baseTransformer.transformTo(prototype, "v", data["type"]);
-            prototype.write("this[clOb].$mangledName = v;},");
-          } else {
-            if (data["name"].startsWith("_")) continue;
-
-            // TODO
-            // (new ClassProperty(data, c, isStatic: true)).render(compiler, functions);
+          if (names.contains(name) || name.startsWith("_")) continue;
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Braces

File: lib/src/compiler/nodes/class.dart [hunk 20]
```diff
@@ -128,96 +192,89 @@ class Class implements Renderable {
...
+          names.add(name);
+
+          // TODO: static fields
+          if (c == null || c.staticFields.contains(name)) continue;
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Braces

File: lib/src/compiler/nodes/class.dart [hunk 21]
```diff
@@ -128,96 +192,89 @@ class Class implements Renderable {
...
+              .split("\n")
+              .where((name) =>
+                  name.length > 0 &&
+                  !name.contains(" ") &&
+                  name.contains(_FIELD_REGEX))
+              .map((name) {
+                try {
+                  if (name.contains(":"))
+                    name = name.substring(name.indexOf(":") + 1);
+                  return _FIELD_REGEX.firstMatch(name).group(0);
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Change FIELD_REGEX to _fieldRegex or _fieldPattern

File: lib/src/compiler/nodes/class.dart [hunk 22]
```diff
@@ -227,9 +284,10 @@ class Class implements Renderable {
 
           var pOutput = new StringBuffer();
           pOutput.write("(");
+          
           (new Func(getters[accessor],
```
Comment (kaendfinger) [inline lib/src/compiler/nodes/class.dart]: Assign to variable.

**Comment by kaendfinger** [comment]
Take my comments with a grain of salt, just some tips for the future more than a fix it all here.

**Comment by Pacane** [comment]
If you want to try the strong mode and the linter, you may add a file called `.analysis_options` with the following content

```yaml
analyzer:
  strong-mode: true

linter:
  rules:
    - camel_case_types
    - empty_constructor_bodies
    - always_declare_return_types
    - camel_case_types
    - constant_identifier_names
    - empty_constructor_bodies
    - implementation_imports
    - library_names
    - library_prefixes
    - non_constant_identifier_names
    - one_member_abstracts
    - package_api_docs
    - package_prefixed_library_names
    - prefer_is_not_empty
    - slash_for_doc_comments
    - super_goes_last
    - type_annotate_public_apis
    - type_init_formals
    - unnecessary_brace_in_string_interp
    - unnecessary_getters_setters
    - avoid_empty_else
    - package_names
    - unrelated_type_equality_checks
    - throw_in_finally
    - close_sinks
    - comment_references
    - control_flow_in_finally
    - empty_statements
    - hash_and_equals
    - iterable_contains_unrelated_type
    - list_remove_unrelated_type
    - test_types_in_equals
    - throw_in_finally
    - valid_regexps
    - annotate_overrides
    - avoid_init_to_null
    - avoid_return_types_on_setters
    - await_only_futures
    - empty_catches
    - prefer_is_not_empty
    - sort_constructors_first
    - sort_unnamed_constructors_first
    - unawaited_futures
```

**Comment by mbullington** [comment]
I'll work on resolving some of the naming issues in future commits that have a larger scope. Thanks though!

**Pull request merged**